Add IntrusiveList utility

rwy0717 opened the pull request:
The IntrusiveList is a doubly linked list, where the node is stored in
the elements itself. The node access mechanism is overridable, allowing
elements to customize node access, or belong in multiple lists
simultaneously.

Compared to the J9_LINKED_LIST, this utility is implemented without
macros while still being type-safe.

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy0717 commented:
Can I get a review @dnakamura or @Leonardo2718 ?

rwy0717 commented:
Addressed some issues with the tests found via offline review with @youngar.

rwy0717 commented:
I should note, CI won't run the intrusive list tests because the OMR_GC_API and OMR_GC_API_TEST flags are disabled. For now, they have to be run manually with something like:

```bash
mkdir build-gc-api && cd build-gc-api
cmake .. -DOMR_GC=ON -DOMR_GC_API=ON -DOMR_GC_API_TEST=ON -DOMR_FVTEST=ON
cmake --build .
ctest -R TestIntrusiveList
```

charliegracie commented on fvtest/gc_api_test/TestIntrusiveList.cpp:
```diff
@@ -0,0 +1,194 @@
+
+using LinkedIntListNode = IntrusiveListNode<LinkedInt>;
+
+static_assert(std::is_standard_layout<LinkedIntListNode>::value,
```
is static assert available on all of our supported compilers?

Leonardo2718 replied:
The compiler component use `statis_assert` so it's available on all supported compilers, presumably.

Leonardo2718 replied:
(see https://github.com/eclipse/omr/blob/97818b41e1a29e687bcfd30a69f04870c343ae2a/compiler/README.md#build-info)

charliegracie replied:
Can you point to files in the compiler component that use it? There is lots of code in the compiler that does not compile on our minimum support compilers.

Leonardo2718 replied:
Sure :slightly_smiling_face: 

https://github.com/eclipse/omr/blob/71f359066d5b6303ae78d81ce4c25c99f4066cff/compiler/env/TRMemory.hpp#L149

https://github.com/eclipse/omr/blob/71f359066d5b6303ae78d81ce4c25c99f4066cff/compiler/il/OMRDataTypes.hpp#L259-L264

https://github.com/eclipse/omr/blob/71f359066d5b6303ae78d81ce4c25c99f4066cff/compiler/il/OMRIL.cpp#L521-L582

https://github.com/eclipse/omr/blob/71f359066d5b6303ae78d81ce4c25c99f4066cff/compiler/il/ILProps.hpp#L126

https://github.com/eclipse/omr/blob/71f359066d5b6303ae78d81ce4c25c99f4066cff/compiler/codegen/OMRCodeGenerator.cpp#L138-L140

There are several more of these and most are in parts of the code that are fundamental to the compiler. :slightly_smiling_face:

charliegracie replied:
perfect! Thanks @Leonardo2718

charliegracie commented on fvtest/gc_api_test/TestIntrusiveList.cpp:
```diff
@@ -0,0 +1,194 @@
+template <>
+struct IntrusiveListNodeAccessor<LinkedInt> {
+
+	static constexpr LinkedIntListNode& node(LinkedInt& i) noexcept { return i.node; }
```
Is constexpr available on xlC and the rest of the supported compilers?

charliegracie commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	// deactivate the node, and clear the next/prev pointers.
+	void clear() noexcept
```
I do not believe noexcept is supported on the minimum version of Visual Studio OMR supports

rwy0717 commented:
I wrote this targeting c++11 but will likely have to make changes to ensure compiler compatibility.

It's for building lists of thread contexts attached to a particular heap, which in turn is needed for coordinating heap access.

rwy0717 commented on fvtest/gc_api_test/TestIntrusiveList.cpp:
```diff
@@ -0,0 +1,194 @@
+};
+
+static_assert(std::is_standard_layout<LinkedInt>::value,
+              "Elements of an intrusive list can be standard layout types.");
```
Not quite--I'm making sure the node won't force the elements to be non-standard-layout. It's still perfectly possible to construct a list of non-standard-layouts elements. I think this comment was written correctly.

Leonardo2718 replied:
Oh, I see. That does make sense. Maybe that explanation can be added somewhere? What that assert is doing is a bit confusing when reading through the code for the first time.

rwy0717 replied:
I will likely have to get rid of the assert anyways, given that our compilers probably don't provide type traits, and I believe it's impossible to implement `is_standard_layout` without compiler support.

rwy0717 commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+
+	/// Add element to the head of the list. Constant time.
+	void add(T* element) noexcept
+	{
```
@charliegracie  I need to get rid of all these asserts below, right? The GC can't use plain old c assert?

rwy0717 commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	IntrusiveListConstIterator<T, Accessor>&
+	operator=(const IntrusiveListConstIterator<T, Accessor>& rhs) noexcept
```
I'm thinking no, given that there are no opportunities to optimize the move constructor. It should be fine to stick with copy assignment (does that sounds right to you?)

Leonardo2718 replied:
Yeah, ok.

To save a few lines of code then, could you default the copy assignment operator, instead of providing a custom definition?

rwy0717 replied:
Honestly I didn't know you could default the assignment operator! Makes sense though. Will do.

rwy0717 commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+
+	explicit constexpr IntrusiveListIterator(T* root) noexcept : _current(root) {}
+
+	constexpr IntrusiveListIterator(const IntrusiveListIterator<T, Accessor>&) noexcept = default;
```
I think providing the copy constructor is fine--there are no special move semantics or optimizations for the iterator.

rwy0717 commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+constexpr bool
+operator!=(const IntrusiveListIterator<T, A>& lhs, const IntrusiveListConstIterator<T, A>& rhs) {
+	return rhs != lhs;
+}
```
My thinking is, in this case, the RHS Iter will implicitly convert to a ConstIter for the comparison, and the compiler will optimize it all to a pointer comparison anyways. There is a test case [here](https://github.com/eclipse/omr/pull/3257/files#diff-028ee6007ebcc7c442f709b9c897840cR62).

rwy0717 replied:
While the non-const iterator is meant to implicitly convert to a const iterator, it does seem a bit unclear to rely on that here. Do you think the few extra lines of code to implement the operators explicitly are worth it? I'm leaning towards yes... or is it just more code for no gain?

how does explicicity

rwy0717 replied:
I added the comparison operators, but decided not to rely on any implicit conversions, to keep things obvious and easy.

youngar commented on fvtest/gc_api_test/TestIntrusiveList.cpp:
```diff
@@ -0,0 +1,194 @@
+{
+
+struct LinkedInt;
+struct LinkedIntTraits;
```
Unused forward declaration
```suggestion
```

youngar commented on gc/api/include/OMR/IntrusiveList.hpp:
```diff
@@ -0,0 +1,308 @@
+class IntrusiveListNode;
+
+template <typename T>
+struct IntrusiveListTraits;
```
Unused

charliegracie commented:
The name of the class seems a little intimidating. Is there are particular reason for this name? Is it a known data structure that I am not familiar with?

I am curious why the omrlinkedlist.h support was not sufficient for your needs? That is the exact code that OMR already uses to create lists of threads for the runtime.

youngar commented:
An "intrusive container" is a fairly well known term for when the object to be stored holds information for the container. I think it is a relatively recent term for an old programming pattern.

rwy0717 commented:
Changes:
* replaced nullptr with NULL (nullptr not available on xlc)
* removed usages of c assert (not allowed in GC code)
* removed noexcept (I think msvc has an issue with this keyword when exceptions are disabled)
* removed static asserts relying on type_traits (is_standard_layout not on xlc)
  * static_assert should be safe to use otherwise
* removed defaulted functions (missing on xlc)
* stopped chaining constructors (just in case)

I left the following 11isms:
* constexpr
* using aliases

rwy0717 commented:
1. How do we feel about the name? Would people rather a simpler/friendlier name, like `List`?

2. Given the overlap in functionality between this code and J9_LINKED_LIST, should I continue pursuing this PR or would it be best just to close it?

3. Currently, this PR is adding the utility to the GC. I'm wondering if this would be better located in include_core, so other components can make use of the code?

rwy0717 commented:
If I move this to include_core, I'm not sure what to do with the unit tests. Would they go in algotest?

rwy0717 commented:
My plan is to move tests to `fvtest/coretest/TestInstrusiveList.cpp`.

`coretest` is a new group of tests for core utilities.

charliegracie commented:
My opinions on this [comment](https://github.com/eclipse/omr/pull/3257#issuecomment-443784606)
1. Name is ok with me.
2. Is it possible to use the already defined macros in this code? That way a change to one is a change to both? I know having an easy to use c++ API is great but we do need to support runtimes that do not use c++.  I am not saying you have to make changes just curious if it is possible since we will now have to support 2 APIs providing the same functionality and the new one will not be easily usable by non C++ runtimes or if there is a wrapper call it will be much slower than using macros.
3. Having this is include_core or util makes sense to me.  I would lean towards util but I am fine with include_core.

rwy0717 commented:
Re 2: I agree duplicate functionality is not ideal. The j9 macros rely on two public fields, linkNext/ linkPrevious. If we put these fields in the element types, and make the fields accessible to the IntrusiveList and Iterator, then we could reuse at least the next/prev/add/remove functionality from the macros.

One potential catch: we `#define linkNext _linkNext` in some sources. I'm imaging that this will not play well with the code in this header. It would definitely violate ODR, and could _maybe_ cause a compilation/link error. If we did this, It would be good to get rid of those defines.

Re 3: Agreed, my first pick for IntrusiveList would have been `util/omrutil`, but omrutil's public headers are stuffed in `include_core` anyways. The only difference would be where testing goes, but given that the util tests need glue, I thought it'd be better to put TestIntrusiveList in the new coretest suite.

rwy0717 commented:
Changes:
* moved header to `include_core/`, tests to `fvtest/coretest/`
* removed using, constexpr
* swapped `<utility>` for `<stddef.h>`

After talking offline with @charliegracie we decided to leave the implementation as-is, rather than use the J9 macros.

rwy0717 commented:
This is ready for a build.

Leonardo2718 commented:
@genie-omr build all

Leonardo2718 commented:
@genie-omr build all

rwy0717 commented:
OK this looks good to go.
